### DISCUSSION TITLE
Passing ERB-templated variables to JS at app load?

### DISCUSSION THREAD

**Discussion by msakrejda** [ElMassimo/vite_ruby#135] [Q&A] [answered]
I've successfully ported our app to Vite (thanks so much for vite_ruby and all your help!), and I'm trying to clean up some crufty corners. Our app has a few places where we pass some variables from ERB/haml to JS functions "installed" (i.e., defined on `window`) by our main entrypoint. The code is essentially something like this (modified from the Rails example), although the real code is more complex and depends on other imports:

```diff
diff --git a/examples/rails/app/frontend/entrypoints/application.ts b/examples/rails/app/frontend/entrypoints/application.ts
index 1b68e3d..1f12af2 100644
--- a/examples/rails/app/frontend/entrypoints/application.ts
+++ b/examples/rails/app/frontend/entrypoints/application.ts
@@ -19,4 +19,6 @@ import.meta.globEager('../channels/**/*_channel.js')
 Turbolinks.start()
 ActiveStorage.start()
 
+window.foo = (arg) => console.log('arg is', arg);
+
 console.log('Visit the guide for more information: ', 'https://vite-ruby.netlify.app/guide/rails')
diff --git a/examples/rails/app/views/layouts/application.html.erb b/examples/rails/app/views/layouts/application.html.erb
index a8d49a3..12a1b79 100644
--- a/examples/rails/app/views/layouts/application.html.erb
+++ b/examples/rails/app/views/layouts/application.html.erb
@@ -11,6 +11,9 @@
 
     <%= vite_stylesheet_tag 'styles.scss' %>
     <%= vite_typescript_tag 'application', 'data-turbo-track': 'reload', media: 'all' %>
+    <script type="text/javascript">
+      window.foo("<%= ENV.fetch('SOME_ENV_VAR', 'default') %>");
+    </script>
   </head>
 
   <body>
```

This doesn't work in vite_ruby because loading the application code happens asynchronously, so the templated, inlined JS runs before the function is even available and fails with `window.foo is not a function`. Right now I'm working around this with

```erb
    <script type="text/javascript">
      let intervalId = setInterval(() => {
        if (foo in window) {
          window.foo("<%= ENV.fetch('SOME_ENV_VAR', 'default') %>");
          clearInterval(intervalId);
        }
      }, 100);
    </script>
```

This works but is obviously pretty gross. Are there better approaches to handling this? Or suggestions for refactoring code like this to work better with Vite? @rollup/plugin-define is not a good fit here, because some of the variables coming from ERB are dynamic.

**Comment by ElMassimo** [comment, marked as answer]
If it's dynamic at startup, you could try using [`vite-plugin-erb`](https://github.com/ElMassimo/vite-plugin-erb). I don't recommend it though, it's slow as it requires calling Ruby from JS.

If the rendered data changes on every request, then the best approach is to ensure the rendered script is also `type="module"`, get a reference to the main script that defines the callback using `vite_asset_path`, and inject an import script.

Pseudo code:

```erb
    <script type="module">
      import { foo } from '<%= vite_asset_path 'application.ts' %>'
      foo("<%= ENV.fetch('SOME_ENV_VAR', 'default') %>");
    </script>
```

Basically, you would leverage the wonderful ES Module system to ensure code is loaded in the correct order. Good bye ordering issues!

Some caveats:
- Module scripts execute only once
- The snippet above assumes native ES modules support (only applies if you are using `plugin-legacy`, else you are already relying on it)

**Reply by msakrejda** [reply]
That's awesome, thanks! Re: "execute only once"--that's just a consideration for dev and HMR, right? Some of these helper functions actually call `ReactDOM.render` with some top-level React components, but that just means the calling code won't be re-executed--any changes to the actual components will still go through HMR?

**Reply by ElMassimo** [reply]
Exactly, and HMR updates are usually performed with a registry, so once the component is mounted, it should work as usual.

**Reply by msakrejda** [reply]
Finally came back to this and got around to implementing this approach. This works great in dev, but in tests with a minified build, I get errors that the exports are not found in the referenced module:

>Uncaught SyntaxError: The requested module '/vite-test/assets/helpers.5deec948.js' does not provide an export named '...'

I guess this is because the template wrappers invoking this code are not taken into account during minification? Reading https://esbuild.github.io/api/#minify-considerations I don't see a great workaround. Any ideas?

**Reply by ElMassimo** [reply]
Try configuring Rollup to [avoid optimizing the exports](https://rollupjs.org/guide/en/#preserveentrysignatures).

```js
preserveEntrySignatures: 'allow-extension',
```

> a minifier usually will not change export signatures

Other relevant options:

[`output.minifyInternalExports`](https://rollupjs.org/guide/en/#outputminifyinternalexports)

**Reply by msakrejda** [reply]
Thanks, that worked great.
